In [11]:
# Import necessary libraries
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.metrics.pairwise import cosine_similarity

customers_df = pd.read_csv("C:/Users/hp/Downloads/Customers.csv")
products_df = pd.read_csv("C:/Users\hp/Downloads/Products.csv")
transactions_df = pd.read_csv("C:/Users/hp/Downloads/Transactions.csv")

In [15]:
customers_df

CustomerID        CustomerName         Region  SignupDate
0        C0001    Lawrence Carroll  South America  2022-07-10
1        C0002      Elizabeth Lutz           Asia  2022-02-13
2        C0003      Michael Rivera  South America  2024-03-07
3        C0004  Kathleen Rodriguez  South America  2022-10-09
4        C0005         Laura Weber           Asia  2022-08-15
..         ...                 ...            ...         ...
195      C0196         Laura Watts         Europe  2022-06-07
196      C0197    Christina Harvey         Europe  2023-03-21
197      C0198         Rebecca Ray         Europe  2022-02-27
198      C0199      Andrea Jenkins         Europe  2022-12-03
199      C0200         Kelly Cross           Asia  2023-06-11

[200 rows x 4 columns]

In [17]:
products_df

ProductID              ProductName     Category   Price
0       P001     ActiveWear Biography        Books  169.30
1       P002    ActiveWear Smartwatch  Electronics  346.30
2       P003  ComfortLiving Biography        Books   44.12
3       P004            BookWorld Rug   Home Decor   95.69
4       P005          TechPro T-Shirt     Clothing  429.31
..       ...                      ...          ...     ...
95      P096     SoundWave Headphones  Electronics  307.47
96      P097       BookWorld Cookbook        Books  319.34
97      P098         SoundWave Laptop  Electronics  299.93
98      P099   SoundWave Mystery Book        Books  354.29
99      P100        HomeSense Sweater     Clothing  126.34

[100 rows x 4 columns]

In [19]:
transactions_df

TransactionID CustomerID ProductID      TransactionDate  Quantity  \
0          T00001      C0199      P067  2024-08-25 12:38:23         1   
1          T00112      C0146      P067  2024-05-27 22:23:54         1   
2          T00166      C0127      P067  2024-04-25 07:38:55         1   
3          T00272      C0087      P067  2024-03-26 22:55:37         2   
4          T00363      C0070      P067  2024-03-21 15:10:10         3   
..            ...        ...       ...                  ...       ...   
995        T00496      C0118      P037  2024-10-24 08:30:27         1   
996        T00759      C0059      P037  2024-06-04 02:15:24         3   
997        T00922      C0018      P037  2024-04-05 13:05:32         4   
998        T00959      C0115      P037  2024-09-29 10:16:02         2   
999        T00992      C0024      P037  2024-04-21 10:52:24         1   

     TotalValue   Price  
0        300.68  300.68  
1        300.68  300.68  
2        300.68  300.68  
3        601.36  300.68  
4        902.04  300.68  
..          ...     ...  
995      459.86  459.86  
996     1379.58  459.86  
997     1839.44  459.86  
998      919.72  459.86  
999      459.86  459.86  

[1000 rows x 7 columns]

In [21]:
# Preprocessing
customers_df['SignupDate'] = pd.to_datetime(customers_df['SignupDate'])
transactions_df['TransactionDate'] = pd.to_datetime(transactions_df['TransactionDate'])

In [23]:
# Merge datasets
transactions_products_df = pd.merge(transactions_df, products_df, on='ProductID', how='left')
complete_data = pd.merge(transactions_products_df, customers_df, on='CustomerID', how='left')

In [59]:
transactions_products_df

TransactionID CustomerID ProductID     TransactionDate  Quantity  \
0          T00001      C0199      P067 2024-08-25 12:38:23         1   
1          T00112      C0146      P067 2024-05-27 22:23:54         1   
2          T00166      C0127      P067 2024-04-25 07:38:55         1   
3          T00272      C0087      P067 2024-03-26 22:55:37         2   
4          T00363      C0070      P067 2024-03-21 15:10:10         3   
..            ...        ...       ...                 ...       ...   
995        T00496      C0118      P037 2024-10-24 08:30:27         1   
996        T00759      C0059      P037 2024-06-04 02:15:24         3   
997        T00922      C0018      P037 2024-04-05 13:05:32         4   
998        T00959      C0115      P037 2024-09-29 10:16:02         2   
999        T00992      C0024      P037 2024-04-21 10:52:24         1   

     TotalValue  Price_x                      ProductName     Category  \
0        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
1        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
2        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
3        601.36   300.68  ComfortLiving Bluetooth Speaker  Electronics   
4        902.04   300.68  ComfortLiving Bluetooth Speaker  Electronics   
..          ...      ...                              ...          ...   
995      459.86   459.86             SoundWave Smartwatch  Electronics   
996     1379.58   459.86             SoundWave Smartwatch  Electronics   
997     1839.44   459.86             SoundWave Smartwatch  Electronics   
998      919.72   459.86             SoundWave Smartwatch  Electronics   
999      459.86   459.86             SoundWave Smartwatch  Electronics   

     Price_y  
0     300.68  
1     300.68  
2     300.68  
3     300.68  
4     300.68  
..       ...  
995   459.86  
996   459.86  
997   459.86  
998   459.86  
999   459.86  

[1000 rows x 10 columns]

In [57]:
complete_data

TransactionID CustomerID ProductID     TransactionDate  Quantity  \
0          T00001      C0199      P067 2024-08-25 12:38:23         1   
1          T00112      C0146      P067 2024-05-27 22:23:54         1   
2          T00166      C0127      P067 2024-04-25 07:38:55         1   
3          T00272      C0087      P067 2024-03-26 22:55:37         2   
4          T00363      C0070      P067 2024-03-21 15:10:10         3   
..            ...        ...       ...                 ...       ...   
995        T00496      C0118      P037 2024-10-24 08:30:27         1   
996        T00759      C0059      P037 2024-06-04 02:15:24         3   
997        T00922      C0018      P037 2024-04-05 13:05:32         4   
998        T00959      C0115      P037 2024-09-29 10:16:02         2   
999        T00992      C0024      P037 2024-04-21 10:52:24         1   

     TotalValue  Price_x                      ProductName     Category  \
0        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
1        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
2        300.68   300.68  ComfortLiving Bluetooth Speaker  Electronics   
3        601.36   300.68  ComfortLiving Bluetooth Speaker  Electronics   
4        902.04   300.68  ComfortLiving Bluetooth Speaker  Electronics   
..          ...      ...                              ...          ...   
995      459.86   459.86             SoundWave Smartwatch  Electronics   
996     1379.58   459.86             SoundWave Smartwatch  Electronics   
997     1839.44   459.86             SoundWave Smartwatch  Electronics   
998      919.72   459.86             SoundWave Smartwatch  Electronics   
999      459.86   459.86             SoundWave Smartwatch  Electronics   

     Price_y          CustomerName         Region SignupDate  
0     300.68        Andrea Jenkins         Europe 2022-12-03  
1     300.68       Brittany Harvey           Asia 2024-09-04  
2     300.68       Kathryn Stevens         Europe 2024-04-04  
3     300.68       Travis Campbell  South America 2024-04-11  
4     300.68         Timothy Perez         Europe 2022-03-15  
..       ...                   ...            ...        ...  
995   459.86            Jacob Holt  South America 2022-01-22  
996   459.86  Mrs. Kimberly Wright  North America 2024-04-07  
997   459.86          Tyler Haynes  North America 2024-09-21  
998   459.86       Joshua Hamilton           Asia 2024-11-11  
999   459.86        Michele Cooley  North America 2024-02-05  

[1000 rows x 13 columns]

In [27]:
# Aggregate customer transaction behavior
customer_features = complete_data.groupby('CustomerID').agg(
    total_spent=('TotalValue', 'sum'),
    total_transactions=('TransactionID', 'nunique'),
    avg_transaction_value=('TotalValue', 'mean'),
    unique_categories=('Category', 'nunique'),
    region=('Region', 'first')
).reset_index()

# Pivot table for product category preferences
category_pivot = pd.pivot_table(
    complete_data, 
    values='Quantity', 
    index='CustomerID', 
    columns='Category', 
    aggfunc='sum', 
    fill_value=0
)

# Combine customer features and product preferences
customer_profiles = pd.merge(customer_features, category_pivot, on='CustomerID', how='left')

In [33]:
customer_features

CustomerID  total_spent  total_transactions  avg_transaction_value  \
0        C0001      3354.52                   5             670.904000   
1        C0002      1862.74                   4             465.685000   
2        C0003      2725.38                   4             681.345000   
3        C0004      5354.88                   8             669.360000   
4        C0005      2034.24                   3             678.080000   
..         ...          ...                 ...                    ...   
194      C0196      4982.88                   4            1245.720000   
195      C0197      1928.65                   3             642.883333   
196      C0198       931.83                   2             465.915000   
197      C0199      1979.28                   4             494.820000   
198      C0200      4758.60                   5             951.720000   

     unique_categories         region  
0                    3  South America  
1                    2           Asia  
2                    3  South America  
3                    3  South America  
4                    2           Asia  
..                 ...            ...  
194                  3         Europe  
195                  2         Europe  
196                  2         Europe  
197                  2         Europe  
198                  4           Asia  

[199 rows x 6 columns]

In [29]:
category_pivot

Category    Books  Clothing  Electronics  Home Decor
CustomerID                                          
C0001           2         0            7           3
C0002           0         4            0           6
C0003           0         4            4           6
C0004           8         0            6           9
C0005           0         0            4           3
...           ...       ...          ...         ...
C0196           3         4            0           5
C0197           0         0            6           3
C0198           0         2            1           0
C0199           0         0            3           6
C0200           4         7            1           4

[199 rows x 4 columns]

In [31]:
customer_profiles

CustomerID  total_spent  total_transactions  avg_transaction_value  \
0        C0001      3354.52                   5             670.904000   
1        C0002      1862.74                   4             465.685000   
2        C0003      2725.38                   4             681.345000   
3        C0004      5354.88                   8             669.360000   
4        C0005      2034.24                   3             678.080000   
..         ...          ...                 ...                    ...   
194      C0196      4982.88                   4            1245.720000   
195      C0197      1928.65                   3             642.883333   
196      C0198       931.83                   2             465.915000   
197      C0199      1979.28                   4             494.820000   
198      C0200      4758.60                   5             951.720000   

     unique_categories         region  Books  Clothing  Electronics  \
0                    3  South America      2         0            7   
1                    2           Asia      0         4            0   
2                    3  South America      0         4            4   
3                    3  South America      8         0            6   
4                    2           Asia      0         0            4   
..                 ...            ...    ...       ...          ...   
194                  3         Europe      3         4            0   
195                  2         Europe      0         0            6   
196                  2         Europe      0         2            1   
197                  2         Europe      0         0            3   
198                  4           Asia      4         7            1   

     Home Decor  
0             3  
1             6  
2             6  
3             9  
4             3  
..          ...  
194           5  
195           3  
196           0  
197           6  
198           4  

[199 rows x 10 columns]

In [35]:
# Normalize numerical features
scaler = StandardScaler()
numeric_features = customer_profiles.iloc[:, 1:].select_dtypes(include=['float64', 'int64'])
normalized_features = scaler.fit_transform(numeric_features)

# Compute similarity matrix
similarity_matrix = cosine_similarity(normalized_features)
similarity_df = pd.DataFrame(similarity_matrix, index=customer_profiles['CustomerID'], columns=customer_profiles['CustomerID'])

In [41]:
# Generate lookalike recommendations for the first 20 customers
lookalikes = {}
for customer_id in customer_profiles['CustomerID'][:20]:
    top_similar = similarity_df[customer_id].nlargest(4)[1:4]  # Exclude self
    lookalikes[customer_id] = list(zip(top_similar.index, top_similar.values))

# Convert lookalikes to DataFrame and save as CSV
lookalike_df = pd.DataFrame.from_dict(lookalikes, orient='index', columns=['SimilarCustomer1', 'SimilarCustomer2', 'SimilarCustomer3'])
lookalike_df.to_csv('C:/Users/hp/Downloads/Lookalike.csv', index_label='CustomerID')

# Output
print("Lookalike recommendations for the first 20 customers have been saved to Lookalike.csv.")

Lookalike recommendations for the first 20 customers have been saved to Lookalike.csv.


The above code generates lookalike recommendations for the first 20 customers based on similarity scores. It goes through each customer and picks top 3 most similar customers by picking the highest similarity values from a similarity matrix excluding the customer itself. It stores these lookalikes in a dictionary, gets the dictionary, and converts it into a DataFrame that gets saved as a CSV file on your computer under the name "Lookalike.csv". Lastly, a success message is echoed to the user that the file is created and saved.

In [61]:
lookalike_df

SimilarCustomer1             SimilarCustomer2  \
C0001  (C0069, 0.8363203472154049)  (C0026, 0.8233826016252753)   
C0002  (C0159, 0.8809297504843957)  (C0062, 0.8798141002033394)   
C0003  (C0166, 0.8157412678250712)  (C0195, 0.7431041137334202)   
C0004  (C0065, 0.9445684642913135)  (C0012, 0.9108381077132794)   
C0005  (C0197, 0.9589496996313328)  (C0095, 0.9402898350962073)   
C0006  (C0196, 0.9073987955256472)  (C0185, 0.9010986449174229)   
C0007  (C0005, 0.9160443981033776)   (C0140, 0.903480468706908)   
C0008  (C0093, 0.9211214263495978)  (C0175, 0.8196075529803727)   
C0009  (C0033, 0.9686691276198753)  (C0083, 0.9675364831160911)   
C0010  (C0077, 0.8756500433208878)  (C0030, 0.8553585230517877)   
C0011  (C0016, 0.8537096242307478)  (C0027, 0.8075309812504569)   
C0012  (C0065, 0.9548401699457965)  (C0004, 0.9108381077132794)   
C0013  (C0099, 0.9262581696808904)   (C0156, 0.902871115398386)   
C0014  (C0128, 0.9902869470464074)  (C0097, 0.9856469301829636)   
C0015  (C0123, 0.9804604457140844)  (C0095, 0.9564576149386843)   
C0016  (C0107, 0.8905180711846299)  (C0011, 0.8537096242307478)   
C0017  (C0075, 0.9397940050740483)  (C0122, 0.8668633434358816)   
C0018  (C0064, 0.8526553273596352)  (C0087, 0.8401394615913937)   
C0019  (C0191, 0.8324349956817032)  (C0174, 0.7363190469983062)   
C0020  (C0140, 0.9523873035976601)   (C0078, 0.946394225046949)   

                  SimilarCustomer3  
C0001  (C0127, 0.7380421876633938)  
C0002  (C0133, 0.8577063793403182)  
C0003  (C0031, 0.7262539455203976)  
C0004  (C0075, 0.8973201423481126)  
C0005  (C0007, 0.9160443981033776)  
C0006  (C0079, 0.7992152736471191)  
C0007  (C0110, 0.8560006459701415)  
C0008  (C0109, 0.8190843552869803)  
C0009  (C0198, 0.9613265449617328)  
C0010  (C0056, 0.8437557825828065)  
C0011  (C0107, 0.7790313852294711)  
C0012  (C0113, 0.8554245663645211)  
C0013  (C0141, 0.8798701862892409)  
C0014  (C0151, 0.9833774402811852)  
C0015  (C0014, 0.9439958845069968)  
C0016  (C0074, 0.7088404382193836)  
C0017  (C0021, 0.8451972892363789)  
C0018  (C0068, 0.8052207942776785)  
C0019  (C0047, 0.7069040423019207)  
C0020  (C0130, 0.9263773917875489)

- (Customer C0001):
Customers Similar To
C0069 (Score: 0.8363)
C0026 (Score: 0.8234)
C0127 (Score: 0.7380)

Description:
Customer C0069 is the one that resembles most with C0001 having a score of ~83.63%.
C0026 and C0127 are less similar with scores at ~82.34% and ~73.80%, respectively.

- (Customer C0007):
Similar Customers:
C0005 (Score: 0.9160)
C0140 (Score: 0.9035)
C0110 (Score: 0.8560)

Description:
Customer C0005 is most alike to C0007 with a high similarity score of ~91.60%.
C0140 and C0110 also have a high similarity score, ~90.35% and ~85.60%, respectively.

- (Customer C0014):
Similar Customers:
C0128 (Similarity Score: 0.9903)
C0097 (Similarity Score: 0.9856)
C0151 (Similarity Score: 0.9834)

- Explanation:
Customer C0014 has very similar behavior to C0128, C0097, and C0151, all of which have similarity scores above ~98%.
This may mean that these customers have nearly identical purchasing patterns or profiles.

In [49]:
from sklearn.metrics.pairwise import pairwise_distances

# Create user-product matrix
user_product_matrix = complete_data.pivot_table(
    index='CustomerID', 
    columns='ProductID', 
    values='Quantity', 
    aggfunc='sum', 
    fill_value=0
)

# Calculate pairwise user similarity
user_similarity = 1 - pairwise_distances(user_product_matrix, metric='cosine')
user_similarity_df = pd.DataFrame(user_similarity, index=user_product_matrix.index, columns=user_product_matrix.index)

# Generate collaborative filtering recommendations
collaborative_lookalikes = {}
for customer_id in user_product_matrix.index[:20]:
    top_similar = user_similarity_df[customer_id].nlargest(4)[1:4]  # Exclude self
    collaborative_lookalikes[customer_id] = list(zip(top_similar.index, top_similar.values))

# Save collaborative filtering recommendations
collaborative_lookalike_df = pd.DataFrame.from_dict(collaborative_lookalikes, orient='index', columns=['SimilarCustomer1', 'SimilarCustomer2', 'SimilarCustomer3'])
collaborative_lookalike_df.to_csv('C:/Users/hp/Downloads/Collaborative_Lookalike.csv', index_label='CustomerID')


This code implements collaborative filtering to generate lookalike recommendations based on customer-product interactions. It starts by building a user-product matrix from the complete dataset, where rows represent customers, and columns represent products, which includes the quantity purchased. Then, it calculates the pairwise cosine similarity between users, which measures how similar customers are based on their purchasing patterns. For the first 20 customers, it finds the top 3 most similar customers excluding the customer itself and stores these lookalikes. The recommendations are saved in a CSV file called "Collaborative_Lookalike.csv" on the user's computer.

In [65]:
collaborative_lookalike_df

SimilarCustomer1              SimilarCustomer2  \
C0001   (C0097, 0.5477225575051661)   (C0194, 0.4696682183138621)   
C0002     (C0030, 0.37282185960072)   (C0091, 0.3339135484518728)   
C0003   (C0134, 0.5199469468957452)   (C0181, 0.5175973113765044)   
C0004   (C0070, 0.4988876515698588)   (C0132, 0.4381780460041328)   
C0005   (C0096, 0.6482037235521645)   (C0055, 0.5144957554275265)   
C0006   (C0058, 0.6488856845230502)    (C0040, 0.629511580291707)   
C0007   (C0020, 0.5883484054145521)  (C0079, 0.49613893835683387)   
C0008   (C0144, 0.3922322702763681)   (C0088, 0.3396831102433788)   
C0009    (C0140, 0.560112033611204)   (C0162, 0.5132649025747366)   
C0010  (C0033, 0.48666426339228763)  (C0077, 0.42479539524287324)   
C0011   (C0135, 0.4841820261350419)  (C0120, 0.35682062007003146)   
C0012  (C0164, 0.45291081365783836)   (C0052, 0.4086031347641993)   
C0013   (C0058, 0.5345224838248488)   (C0040, 0.5185629788417315)   
C0014   (C0128, 0.8944271909999159)   (C0159, 0.7844645405527362)   
C0015   (C0073, 0.6324555320336758)   (C0036, 0.6067798762169179)   
C0016   (C0187, 0.6449020216370241)   (C0064, 0.4611123341633879)   
C0017  (C0037, 0.39886201760873274)   (C0071, 0.3849001794597504)   
C0018  (C0177, 0.39389277113386467)    (C0064, 0.356654251494682)   
C0019   (C0083, 0.4898979485566357)   (C0064, 0.4879954250643349)   
C0020   (C0026, 0.6172133998483676)   (C0007, 0.5883484054145521)   

                   SimilarCustomer3  
C0001   (C0199, 0.4381780460041329)  
C0002    (C0071, 0.329914439536929)  
C0003   (C0144, 0.3999999999999999)  
C0004   (C0063, 0.3360672201667223)  
C0005   (C0064, 0.3328770246548891)  
C0006  (C0178, 0.37463432463267754)  
C0007  (C0026, 0.36313651960128146)  
C0008  (C0003, 0.31378581622109447)  
C0009   (C0062, 0.4950737714883371)  
C0010  (C0094, 0.40590605489504306)  
C0011  (C0173, 0.33075929223788925)  
C0012   (C0173, 0.3577708763999663)  
C0013   (C0169, 0.4409585518440985)  
C0014  (C0136, 0.46188021535170054)  
C0015   (C0059, 0.4568321925761285)  
C0016   (C0145, 0.3954372976473721)  
C0017  (C0183, 0.34570865256115424)  
C0018   (C0081, 0.3503245248726854)  
C0019     (C0039, 0.46475800154489)  
C0020   (C0104, 0.4886777774252209)

In [78]:
# Load all results
cosine_lookalikes = pd.read_csv('C:/Users/hp/Downloads/Lookalike.csv', index_col='CustomerID')
collaborative_lookalikes = pd.read_csv('C:/Users/hp/Downloads/Collaborative_Lookalike.csv', index_col='CustomerID')

# Combine into a single DataFrame for comparison
comparison_df = pd.concat(
    [cosine_lookalikes, collaborative_lookalikes],
    keys=['Cosine Similarity', 'Collaborative Filtering'],
    axis=1
)

# Save combined results
comparison_df.to_csv('C:/Users/hp/Downloads/Model_Comparison.csv')

In [80]:
comparison_df

Cosine Similarity                                 \
                         SimilarCustomer1               SimilarCustomer2   
CustomerID                                                                 
C0001       ('C0069', 0.8363203472154049)  ('C0026', 0.8233826016252753)   
C0002       ('C0159', 0.8809297504843957)  ('C0062', 0.8798141002033394)   
C0003       ('C0166', 0.8157412678250712)  ('C0195', 0.7431041137334202)   
C0004       ('C0065', 0.9445684642913135)  ('C0012', 0.9108381077132794)   
C0005       ('C0197', 0.9589496996313328)  ('C0095', 0.9402898350962073)   
C0006       ('C0196', 0.9073987955256472)  ('C0185', 0.9010986449174229)   
C0007       ('C0005', 0.9160443981033776)   ('C0140', 0.903480468706908)   
C0008       ('C0093', 0.9211214263495978)  ('C0175', 0.8196075529803727)   
C0009       ('C0033', 0.9686691276198753)  ('C0083', 0.9675364831160911)   
C0010       ('C0077', 0.8756500433208878)  ('C0030', 0.8553585230517877)   
C0011       ('C0016', 0.8537096242307478)  ('C0027', 0.8075309812504569)   
C0012       ('C0065', 0.9548401699457965)  ('C0004', 0.9108381077132794)   
C0013       ('C0099', 0.9262581696808904)   ('C0156', 0.902871115398386)   
C0014       ('C0128', 0.9902869470464074)  ('C0097', 0.9856469301829636)   
C0015       ('C0123', 0.9804604457140844)  ('C0095', 0.9564576149386843)   
C0016       ('C0107', 0.8905180711846299)  ('C0011', 0.8537096242307478)   
C0017       ('C0075', 0.9397940050740483)  ('C0122', 0.8668633434358816)   
C0018       ('C0064', 0.8526553273596352)  ('C0087', 0.8401394615913937)   
C0019       ('C0191', 0.8324349956817032)  ('C0174', 0.7363190469983062)   
C0020       ('C0140', 0.9523873035976601)   ('C0078', 0.946394225046949)   

                                                  Collaborative Filtering  \
                         SimilarCustomer3                SimilarCustomer1   
CustomerID                                                                  
C0001       ('C0127', 0.7380421876633938)   ('C0097', 0.5477225575051661)   
C0002       ('C0133', 0.8577063793403182)     ('C0030', 0.37282185960072)   
C0003       ('C0031', 0.7262539455203976)   ('C0134', 0.5199469468957452)   
C0004       ('C0075', 0.8973201423481126)   ('C0070', 0.4988876515698588)   
C0005       ('C0007', 0.9160443981033776)   ('C0096', 0.6482037235521645)   
C0006       ('C0079', 0.7992152736471191)   ('C0058', 0.6488856845230502)   
C0007       ('C0110', 0.8560006459701415)   ('C0020', 0.5883484054145521)   
C0008       ('C0109', 0.8190843552869803)   ('C0144', 0.3922322702763681)   
C0009       ('C0198', 0.9613265449617328)    ('C0140', 0.560112033611204)   
C0010       ('C0056', 0.8437557825828065)  ('C0033', 0.48666426339228763)   
C0011       ('C0107', 0.7790313852294711)   ('C0135', 0.4841820261350419)   
C0012       ('C0113', 0.8554245663645211)  ('C0164', 0.45291081365783836)   
C0013       ('C0141', 0.8798701862892409)   ('C0058', 0.5345224838248488)   
C0014       ('C0151', 0.9833774402811852)   ('C0128', 0.8944271909999159)   
C0015       ('C0014', 0.9439958845069968)   ('C0073', 0.6324555320336758)   
C0016       ('C0074', 0.7088404382193836)   ('C0187', 0.6449020216370241)   
C0017       ('C0021', 0.8451972892363789)  ('C0037', 0.39886201760873274)   
C0018       ('C0068', 0.8052207942776785)  ('C0177', 0.39389277113386467)   
C0019       ('C0047', 0.7069040423019207)   ('C0083', 0.4898979485566357)   
C0020       ('C0130', 0.9263773917875489)   ('C0026', 0.6172133998483676)   

                                                                            
                          SimilarCustomer2                SimilarCustomer3  
CustomerID                                                                  
C0001        ('C0194', 0.4696682183138621)   ('C0199', 0.4381780460041329)  
C0002        ('C0091', 0.3339135484518728)    ('C0071', 0.329914439536929)  
C0003        ('C0181', 0.5175973113765044)   ('C0144', 0.3999999999999999)  
C0004        ('C0132', 0.4381780460041328

There are two recommendation techniques, namely Cosine Similarity and Collaborative Filtering. Both exhibit different focuses and strengths in identifying similarities among customers. Cosine Similarity consistently identifies customers with a high similarity score based on general patterns of behavior or shared characteristics. In comparison, Collaborative Filtering provides recommendations based on the interactions of the users and the items or historical activity. This approach often brings out more divergent suggestions of lower similarity score, emphasizing a unique aspect that customer preferences and the transaction history exhibit and are unable to capture via Cosine Similarity.

Despite their differences, these methods complement each other very well. Cosine Similarity is very good at finding customers with similar features or who behave alike, while Collaborative Filtering reveals subtle relationships through past interactions. Combining these techniques provides a more complete understanding of customer behavior, allowing businesses to identify both broad patterns and subtle connections that may be missed when relying on a single method.